New component 'insteon_plm' and related platforms

nugget opened the pull request:
**Description:**

This PR adds a new component 'insteon_plm' for use with INSTEON PowerLinc USB/serial modems. It includes constituent platforms: light, switch, and binary_sensor.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#2092

**Example entry for `configuration.yaml` (if applicable):**
```yaml
insteon_plm:
    port: /dev/ttyUSB0
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51

nugget committed:
Connect to PLM and process simple protocol callbacks

nugget committed:
Connect to PLM and process simple protocol callbacks

nugget committed:
Baseline commit

nugget committed:
Baseline commit

nugget committed:
Merge branch 'insteon_plm' of github.com:nugget/home-assistant into insteon_plm

nugget committed:
Connection working again

nugget committed:
Async add devices is working via callback now

nugget committed:
Beginning to interface with PLM library for control and state

nugget committed:
Deal with brightness in 255 levels with library

nugget committed:
Change sub names to match API changes

nugget committed:
Remove PLM-level update callback

nugget committed:
Support dimmable based on underlying PLM device attributes

nugget committed:
Expand to non-light platforms

nugget committed:
Stubs for turn on and off

nugget committed:
Current version of Python library

nugget committed:
Amend to use switch device attributes

nugget committed:
Use asyncio endpoints for control

nugget committed:
Add logging line

nugget committed:
Bump module version to 0.7.1

nugget committed:
Auto-load platforms, display device info/attributes

nugget committed:
Unify method name for getting a device attribute

nugget committed:
Require Current version of insteonplm module

nugget committed:
Import the component function in each platform in the balloob-recommend manner

nugget committed:
For consistency, handle switch state as onlevel just like lights

nugget committed:
Use level 0xff for on state, even with binary switches

Observing the behavior of a 2477S switch, it looks like even the non-dimmable
devices use 0x00 and 0xff for off/on respectively.  I was using 0x01 for on
previously, but that yields unnecessary state change callbacks when message
traffic ends up flipping the onlevel from 0xff to 0x01 or 0x01 to 0xff.

nugget committed:
Use sensorstate attribute for sensor onoff

nugget committed:
Move new device callback to devices attribute

nugget committed:
Add support for platform override on a device

nugget committed:
Bump version of insteonplm module

nugget committed:
Default overrides is an empty list

nugget committed:
Avoid calling private methods when doing common attributes

nugget committed:
Remove unused CONF_DEBUG for now

nugget committed:
flake8 and pylint code cleanup

nugget committed:
Move get_component to local function where it is needed

nugget committed:
Update to include insteonplm module.

nugget committed:
New files for insteon_plm component

nugget committed:
Legitimate class doctring instead of stub

nugget committed:
Docstring changes.

SEJeff commented on homeassistant/components/binary_sensor/insteon_plm.py:
```diff
@@ -0,0 +1,96 @@
+import logging
+import asyncio
+
+from homeassistant.components.binary_sensor import (BinarySensorDevice)
```
For a single import, the `()` are unnecessary.

SEJeff commented on homeassistant/components/binary_sensor/insteon_plm.py:
```diff
@@ -0,0 +1,96 @@
+        hass.async_add_job(async_add_devices(
+            [InsteonPLMBinarySensorDevice(hass, plm, address, name)]))
+
+    criteria = dict(capability='binary_sensor')
```
Totally minor stylistic issue, but would you care to use literals here ie:

    criteria = {'capability': 'binary_sensor'}

Besides the fact that it doesn't require a function call, I (personally) find it more obvious.

SEJeff commented on homeassistant/components/insteon_plm.py:
```diff
@@ -0,0 +1,96 @@
+    plm = yield from insteonplm.Connection.create(
+        device=port, loop=hass.loop)
+
+    print(overrides)
```
If you'd like this to be printed to the log, would you mind using something more like:

    _LOGGER.debug('Overrides: %s', overrides)

SEJeff commented on homeassistant/components/insteon_plm.py:
```diff
@@ -0,0 +1,96 @@
+    attributekeys['cat'] = 'Category'
+    attributekeys['subcat'] = 'Subcategory'
+    attributekeys['firmware'] = 'Firmware'
+    attributekeys['product_key'] = 'Product Key'
```
Simple stylistic suggestion to make this a bit more pleasant:

```python
attributekeys = {
    'address': 'INSTEON Address',
    'description': 'Description',
    'model': 'Model',
    'cat': 'Category',
    'subcat': 'Subcategory',
    'firmware': 'Firmware',
    'product_key': 'Product Key',
}
```

Achieves the same functionality without duplicating `attributekeys['` over and over

SEJeff commented on homeassistant/components/light/insteon_plm.py:
```diff
@@ -0,0 +1,121 @@
+        hass.async_add_job(async_add_devices(
+            [InsteonPLMDimmerDevice(hass, plm, address, name, dimmable)]))
+
+    criteria = dict(capability='light')
```
Same as previously. Would you mind using dict literals here please?

SEJeff commented on homeassistant/components/light/insteon_plm.py:
```diff
@@ -0,0 +1,121 @@
+        self._dimmable = dimmable
+
+        self._plm.add_update_callback(
+            self.async_light_update, dict(address=self._address))
```
pretty please can you not do this? :)

SEJeff commented on homeassistant/components/light/insteon_plm.py:
```diff
@@ -0,0 +1,121 @@
+        if ATTR_BRIGHTNESS in kwargs:
+            brightness = int(kwargs[ATTR_BRIGHTNESS])
+        else:
+            brightness = 255
```
Totally sensible, but any chance you'd consider making this a module level constant? [Magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)#Unnamed_numerical_constants) are always a PITA even though your code is super straightforward.

SEJeff commented on homeassistant/components/switch/insteon_plm.py:
```diff
@@ -0,0 +1,101 @@
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the INSTEON PLM device class for the hass platform."""
+    _LOGGER.info('Provisioning Insteon PLM Switches')
```
Stupid question, but would it make sense to differentiate this log message for each of the new Insteon PML components?

nugget replied:
Thanks for all the feedback, I incorporated your other suggestions, but this one is puzzling me.  I don't really understand what you're asking.  Each of the three platforms has an equivalent of this log line, but they are differentiated (specifically referencing "Switches" and "Lights" and "Binary Sensors."

SEJeff replied:
I'm clearly high on all of the cough syrup I took (I'm sick as a dog right now and decided to review some code to ease my mind). You're entirely right, so please ignore this.

nugget replied:
Instructions unclear, added "import DayQuil" to class.

SEJeff reviewed (commented):
Really solid stuff, but would you please consider changing all of the `dict()` invocations to the type literal invocations ie: `{}` please?

If you want to nerd out, the `dict()` calls the function constructor while the more literal `{}` just calls the [BUILD_MAP](https://docs.python.org/3/library/dis.html#opcode-BUILD_MAP) bytecode directly:

See for yourself in the interpreter using the `dis` module:

```python

>>> dis.dis('dict(foo=1)')
  1           0 LOAD_NAME                0 (dict)
              3 LOAD_CONST               0 ('foo')
              6 LOAD_CONST               1 (1)
              9 CALL_FUNCTION          256 (0 positional, 1 keyword pair)
             12 RETURN_VALUE
>>> dis.dis('{"foo": 1}')
  1           0 LOAD_CONST               0 ('foo')
              3 LOAD_CONST               1 (1)
              6 BUILD_MAP                1
              9 RETURN_VALUE
>>> 
```

nugget committed:
Style changes as requested by @SEJeff

pvizeli commented on homeassistant/components/binary_sensor/insteon_plm.py:
```diff
@@ -0,0 +1,96 @@
+        async_plm_binarysensor_callback, criteria)
+
+    new_binarysensors = []
+    yield from async_add_devices(new_binarysensors)
```
don't call this with none device...

pvizeli commented on homeassistant/components/binary_sensor/insteon_plm.py:
```diff
@@ -0,0 +1,96 @@
+    @property
+    def is_on(self):
+        """Return the boolean response if the node is on."""
+        sensorstate = self._plm.get_device_attr(self._address, 'sensorstate')
```
Don't make IO on property

nugget replied:
This method does not cause any IO, it's just probing a private data structure within the insteon_plm class ([ref](https://github.com/nugget/python-insteonplm/blob/master/insteonplm/protocol.py#L815))

pvizeli commented on homeassistant/components/binary_sensor/insteon_plm.py:
```diff
@@ -0,0 +1,96 @@
+    @property
+    def sensor_class(self):
+        """Sensor class is unknown."""
+        return
```
Remove it, None is default

pvizeli commented on homeassistant/components/binary_sensor/insteon_plm.py:
```diff
@@ -0,0 +1,96 @@
+        _LOGGER.info('New INSTEON PLM binary sensor device: %s (%s)',
+                     name, address)
+        hass.async_add_job(async_add_devices(
+            [InsteonPLMBinarySensorDevice(hass, plm, address, name)]))
```
You not need the `async_add_devices` you can create your entity and call `entity.async_update_ha_state`

pvizeli commented on homeassistant/components/binary_sensor/insteon_plm.py:
```diff
@@ -0,0 +1,96 @@
+    def async_binarysensor_update(self, message):
+        """Receive notification from transport that new data exists."""
+        _LOGGER.info('Received update calback from PLM for %s', self._address)
+        self._hass.async_add_job(self.async_update_ha_state(True))
```
don't call this with `True`, you have no update function.

pvizeli commented on homeassistant/components/light/insteon_plm.py:
```diff
@@ -0,0 +1,123 @@
+
+        _LOGGER.info('New INSTEON PLM light device: %s (%s)', name, address)
+        hass.async_add_job(async_add_devices(
+            [InsteonPLMDimmerDevice(hass, plm, address, name, dimmable)]))
```
same too

pvizeli commented on homeassistant/components/light/insteon_plm.py:
```diff
@@ -0,0 +1,123 @@
+        async_plm_light_callback, criteria)
+
+    new_lights = []
+    yield from async_add_devices(new_lights)
```
same too

pvizeli commented on homeassistant/components/light/insteon_plm.py:
```diff
@@ -0,0 +1,123 @@
+    def async_light_update(self, message):
+        """Receive notification from transport that new data exists."""
+        _LOGGER.info('Received update calback from PLM for %s', self._address)
+        self._hass.async_add_job(self.async_update_ha_state(True))
```
Same too

pvizeli commented on homeassistant/components/switch/insteon_plm.py:
```diff
@@ -0,0 +1,101 @@
+        _LOGGER.info('New INSTEON PLM switch device: %s (%s)', name, address)
+        hass.async_add_job(
+            async_add_devices(
+                [InsteonPLMSwitchDevice(hass, plm, address, name)]))
```
same too

pvizeli commented on homeassistant/components/switch/insteon_plm.py:
```diff
@@ -0,0 +1,101 @@
+        async_plm_switch_callback, criteria)
+
+    new_switches = []
+    yield from async_add_devices(new_switches)
```
same too

pvizeli reviewed (changes requested):
Please make no IO inside propertys

nugget committed:
Changes requested by @pvizeli

pvizeli commented:
Can you register a callback on component and use discovery for setup platforms? You way overwrite the idea of discovery. You can use discovery info for get data to it.

balloob commented on homeassistant/components/light/insteon_plm.py:
```diff
@@ -0,0 +1,120 @@
+
+    plm = hass.data['insteon_plm']
+
+    def async_plm_light_callback(device):
```
Can you annotate this with `@callback` (from `homeassistant.core`). It improves readability.

nugget committed:
Add @callback decorator to callback functions

nugget commented:
@balloob , @pvizeli : After our conversation on Gitter last night, I messed around deferred load_platform suggestion and got some code in place that *functions* but it's so terribly ugly and unwieldy that I don't even want to commit it. 

I grepped the dev branch to try to find some other component that behaves this way and I came up empty.  Everywhere I look I just see components doing a "for p in platform_list" and loading all their platforms at once during setup.  The only component I found that did anything remotely like this is hdmi_cec but it's pretty bare-bones and didn't really help to educate me about the intent of the various discovery helpers.  Do you have an example I could look at to guide my development?

The code [at this gist](https://gist.github.com/nugget/b2abaac848110e75b4efd55956379e9b) functions but it's truly awful (even ignoring all the print() call breadcrumbs.

I'm happy to pursue this, it makes sense especially since the number of insteon_plm platforms is only going to grow in the future (definitely adding sensor, probably cover and lock as well).  But from the trenches here it's hard to see how the complexity is worth the tradeoff of just loading potentially unused platforms for insteon_plm.  That tells me I'm probably missing something fundamental about the suggestion.

Thanks!

pvizeli commented:
It is not very dificult. You need to have callback and that generate a config they will be discovered. On platform you need only to look at discovery_info and Setup device with that.

Do you have allready the code for that, move that to component and send the `address`, `address_hex` as config to platform.

nugget commented:
Oh, ok, I think I see the intent.  Thanks

pvizeli commented:
On platform
```python
@asyncio.coroutine
def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
    """Set up the INSTEON PLM device class for the hass platform."""
    if discovery_info is None:
        return
    _LOGGER.info('Provisioning Insteon PLM Binary Sensors')

    plm = hass.data['insteon_plm']

    name = discovery_info['address']
    address = discovery_info['address_hex']
    yield from async_add_devices([InsteonPLMBinarySensorDevice(hass, plm, address, name)]))
```

nugget committed:
Opportunistic platform loading triggered by qualifying device detection

Instead of loading all the constituent platforms that comprise the insteon_plm
component, instead we defer and wait until we receive a callback for a device
that requires the platform.

nugget commented:
OK, I think I got it.  Committed and pushed changes which move the new device callback handling to the component level and then pass the new devices down to the relevant platform(s) for that device using the discovery_info channel of async_load_platform.   Thanks again for the help and guidance.  Every time I submit code to this project I learn and improve.

pvizeli reviewed (approved):
Amazing work :zap:
